src/main/java/ru/vidtu/ias/auth/microsoft/MSAuth.java

```diff
@@ -0,0 +1,564 @@
+                "&refresh_token=" + URLEncoder.encode(refresh, StandardCharsets.UTF_8) +
+                "&grant_type=refresh_token" +
+                "&scope=XboxLive.signin%20XboxLive.offline_access";
+
+        // Send the request.
+        return CLIENT.sendAsync(HttpRequest.newBuilder()
+                .uri(createURIUnchecked("https://login.live.com/oauth20_token.srf"))
+                .header("User-Agent", IAS.userAgent())
+                .header("Accept", "application/json")
+                .header("Content-Type", "application/x-www-form-urlencoded")
+                .timeout(IAS.TIMEOUT)
+                .POST(HttpRequest.BodyPublishers.ofString(payload))
+                .build(), HttpResponse.BodyHandlers.ofString()).thenApplyAsync(response -> {
+            // Process the response.
+            try {
+                // Check the code.
+                int status = response.statusCode();
+                if (status != HttpURLConnection.HTTP_OK) {
+                    throw new IllegalArgumentException("Invalid status code: " + status);
+                }
+
+                // Decode the tokens and return them.
+                JsonObject json = GSONUtils.GSON.fromJson(response.body(), JsonObject.class);
+                Objects.requireNonNull(json, "Response is null");
+                return MSTokens.fromJson(json);
+            } catch (Throwable t) {
+                // Rethrow, trying to remove sensitive data.
+                String message = "Unable to convert Microsoft Refresh (MSR) token to Microsoft Access (MSA) and Microsoft Refresh (MSR) tokens (" + response + " with " + response.headers() + "): " + response.body();
+                message = message.replace(refresh, "[MSR]");
+                throw new RuntimeException(message, t);
+            }
+        }, IAS.executor());
+    }
+
+    /**
+     * Gets the Xbox Live (XBL) token from the Microsoft Access (MSA) token.
+     *
+     * @param authToken Microsoft Access (MSA) token (e.g. from {@link MSTokens#access()})
+     * @return Future that will complete with an XBL token and a user hash or exceptionally
+     * @see <a href="https://wiki.vg/Microsoft_Authentication_Scheme#Authenticate_with_Xbox_Live">Reference</a>
+     */
+    public static CompletableFuture<XHashedToken> msaToXbl(String authToken) {
+        // Create the payload.
+        // This is ugly, but I won't create custom classes and serializers just for this.
+        JsonObject request = new JsonObject();
+
+        JsonObject requestProperties = new JsonObject();
+        requestProperties.addProperty("AuthMethod", "RPS");
+        requestProperties.addProperty("SiteName", "user.auth.xboxlive.com");
+        requestProperties.addProperty("RpsTicket", "d=" + authToken);
+        request.add("Properties", requestProperties);
+
+        // We disable "HTTP -> HTTPS" inspection here, because it's not an actual URL,
+        // but a payload parameter (possibly) required by the specification:
```

Comment:
Have you tested with https in the parameter to see if it works? I know the spec shows an http URL but it doesn't say there that it _has_ to be http

Reply:
yes, it return 400 with HTTPS. not a critical issue, since it's parameter name:
```
Caused by: java.lang.IllegalArgumentException: Invalid status code: 400
	at ru.vidtu.ias.auth.microsoft.MSAuth.lambda$msaToXbl$3(MSAuth.java:326) ~[main/:?]
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646) ~[?:?]
	... 7 more
```